[FLINK-12331] Introduce partition/gate setup to decouple task registration with NetworkEnvironment

azagrebin opened the pull request:
## What is the purpose of the change

Introduce partition/gate setup to decouple task registration with NetworkEnvironment.
The PR also does some preparation refactoring.

## Brief change log

  - Introduce factories for ResultPartion and SingleInputgates
  - Introduce buffer pool factories for ResultPartion and SingleInputgates
  - Introduce MemorySegmentProvider for RemoteInputChannel to assign exclusive segments
  - Refactor NetworkEnvironment#setupXXX() to ResultPartitionWriter#setup() and InputGate#setup()

## Verifying this change

existing unit tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

azagrebin commented:
@flinkbot attention @zhijiangW

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java:
```diff
@@ -98,15 +99,15 @@ private NetworkEnvironment(
 		this.taskEventPublisher = taskEventPublisher;
-		this.ioManager = ioManager;
 		this.isShutdown = false;
+		this.resultPartitionFactory = resultPartitionFactory;
```
`this.resultPartitionFactory = resultPartitionFactory` in front of `this.isShutdown = false`?

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java:
```diff
@@ -108,6 +112,7 @@ private NetworkEnvironment(
 		this.taskEventPublisher = taskEventPublisher;
 		this.isShutdown = false;
```
Put `isShutdown=false` at the end of this constructor?

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java:
```diff
@@ -128,13 +133,17 @@ public static NetworkEnvironment create(
 			new ResultPartitionFactory(resultPartitionManager, checkNotNull(ioManager));
 
+		SingleInputGateFactory singleInputGateFactory =
+			new SingleInputGateFactory(config, connectionManager, resultPartitionManager);
```
`TaskEventDispatcher` could also be included in `singleInputGateFactory`?

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGateTest.java:
```diff
@@ -415,16 +395,19 @@ public void testRequestBuffersWithRemoteInputChannel() throws Exception {
+				InputChannelBuilder.newBuilder()
+					.setupFromNetworkEnvironment(network)
+					.buildUnknown(inputGate)
+					.toRemoteInputChannel(InputChannelBuilder.STUB_CONNECTION_ID);
```
why not build remote directly?

azagrebin replied:
I tried to preserve the previous way of building remote but looking at it more, true, it does not seem to be relevant.

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGateTest.java:
```diff
@@ -415,16 +395,19 @@ public void testRequestBuffersWithRemoteInputChannel() throws Exception {
+					.buildUnknown(inputGate)
+					.toRemoteInputChannel(InputChannelBuilder.STUB_CONNECTION_ID);
+			IntermediateResultPartitionID resultPartitionId = remoteFromUnknown.getPartitionId().getPartitionId();
+			inputGate.setInputChannel(resultPartitionId, remoteFromUnknown);
```
`setInputChannel` could be removed?

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGateTest.java:
```diff
@@ -415,16 +395,19 @@ public void testRequestBuffersWithRemoteInputChannel() throws Exception {
 			if (enableCreditBasedFlowControl) {
-				RemoteInputChannel remote = (RemoteInputChannel) inputGate.getInputChannels()
-					.get(resultPartitionId.getPartitionId());
+				RemoteInputChannel remote = (RemoteInputChannel) inputGate.getInputChannels().get(resultPartitionId);
```
Use the previous `remoteFromUnknown` directly, so `IntermediateResultPartitionID resultPartitionId = remoteFromUnknown.getPartitionId().getPartitionId()` could also be removed.

zhijiangW commented on flink-core/src/main/java/org/apache/flink/core/memory/MemorySegmentProvider.java:
```diff
@@ -0,0 +1,34 @@
+	/** Requests default number of memory segments. */
+	Collection<MemorySegment> requestMemorySegments() throws IOException;
+
+	Collection<MemorySegment> requestMemorySegments(int numRequiredBuffers) throws IOException;
```
If we only consider the semantics of `MemorySegmentProvider` interface, the `requestMemorySegments()` is a bit redundant, which could be replaced by `requestMemorySegments(int numRequiredBuffers)`. The only reason we keep both methods is that it seems not elegant to maintain `networkBuffersPerChannel` in `SingleInputGate`.

TBH, I do not think it is worth defining this new interface not very clean to accommodate the implementation. Because `requestMemorySegments()` seems blurry for the usage which relies on the default value, and we should also consider if the default value is 0, whether the returned collection is null or empty size. 
In addition, `requestMemorySegments(int numRequiredBuffers)` is only used for tests ATM.

azagrebin replied:
I think `MemorySegmentProvider` makes sense from point of interface segregation principle. This interface encapsulates segment management concern which is the only thing that remote channel needs atm. `NetworkBufferPool` looks like implementing 2 separate concerns: `BufferPoolFactory` and `MemorySegmentProvider`.

As for `requestMemorySegments(int numRequiredBuffers)`, you are right, it is more clear to remove it from the interface and adjust the tests with explicit default value, e.g. 1.

zhijiangW replied:
We are indeed in the lack of semantics of requesting/releasing batch of buffers before, and I think it is reasonable to extend these methods. 

- `SegmentProvider` is not really needed because `RemoteInputChannel` still needs construct the `Buffer` internally after requesting `MemorySegment`.

- We could extend the general `BufferProvider` interface to provide method of `requestBuffers(int numRequiredBuffers, BufferRecycler)`,  and `RemoteInputChannel` already implements `BufferRecycler` interface.

- We could also extend the general `BufferRecycler`  interface to provide method of `recycle(Collection<MemorySegment>)`. 

To do so we make use of the existing semantics and architecture. But the only concern is we would affect the previous implementations to bring more changes.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGate.java:
```diff
@@ -155,8 +155,8 @@
 
-	/** Global network buffer pool to request and recycle exclusive buffers (only for credit-based). */
-	private NetworkBufferPool networkBufferPool;
+	/** Global global memory segment provider to request and recycle exclusive buffers (only for credit-based). */
```
repeated `global`

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGate.java:
```diff
@@ -535,7 +510,7 @@ public void requestPartitions() throws IOException, InterruptedException {
 		while (true) {
 			synchronized (inputChannelsWithData) {
-				while (inputChannelsWithData.size() == 0) {
+				while (inputChannelsWithData.isEmpty()) {
```
why make this change?

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGateFactory.java:
```diff
@@ -121,8 +127,8 @@ else if (partitionLocation.isRemote()) {
-					metrics
-				);
+					metrics,
+					memorySegmentProvider);
```
Whether to separate line for last `);` which keeps the consistent with `local` and `unknown` cases?

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/netty/PartitionRequestClientHandlerTest.java:
```diff
@@ -203,14 +205,20 @@ public void testCancelBeforeActive() throws Exception {
+	static RemoteInputChannel createRemoteInputChannel(
+		SingleInputGate inputGate, MemorySegmentProvider memorySegmentProvider) throws Exception {
+
+		return createRemoteInputChannel(inputGate, mock(PartitionRequestClient.class), 0, 0, memorySegmentProvider);
```
It is better to avoid mock `PartitionRequestClient` here.

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/netty/PartitionRequestClientHandlerTest.java:
```diff
@@ -237,7 +254,8 @@ static RemoteInputChannel createRemoteInputChannel(
+			int maxBackoff,
+			MemorySegmentProvider memorySegmentProvider) throws Exception {
 		final ConnectionManager connectionManager = mock(ConnectionManager.class);
 		when(connectionManager.createPartitionRequestClient(any(ConnectionID.class)))
```
There are already too many utils here to construct the `RemoteInputChannel`, but we still miss the units of `{SingleInputGate, PartitionRequestClient, MemorySegmentProvider}` which could avoid pass `createRemoteInputChannel(inputGate, client, 0, 0, networkBufferPool)` in `CreditBasedPartitionRequestClientHandlerTest`. 

Maybe we need a separate hotfix for refactoring these parts to make `connectionManager.createPartitionRequestClient` call in separate tests if necessary, then we only need one type to construct `RemoteInputChannel`.

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/InputChannelBuilder.java:
```diff
@@ -46,6 +48,7 @@
 	private int maxBackoff = 0;
 	private InputChannelMetrics metrics =
 		new InputChannelMetrics(UnregisteredMetricGroups.createUnregisteredTaskMetricGroup());
```
Here we could use `InputChannelTestUtils.newUnregisteredInputChannelMetrics`

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannelTest.java:
```diff
@@ -1010,8 +1032,9 @@ private RemoteInputChannel createRemoteInputChannel(
 		return InputChannelBuilder.newBuilder()
 			.setConnectionManager(connectionManager)
-			.setInitialBackoff(initialBackoff)
+			.setInitialBackoff(initialBackOff)
```
No need change?

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannelTest.java:
```diff
@@ -992,16 +993,37 @@ public void testConcurrentRecycleAndRelease2() throws Exception {
+	private RemoteInputChannel createRemoteInputChannel(
+		SingleInputGate inputGate,
+		PartitionRequestClient partitionRequestClient,
+		int initialBackOff,
```
lowercase `off` to keep consistent with maxBackoff?

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannelTest.java:
```diff
@@ -992,16 +993,37 @@ public void testConcurrentRecycleAndRelease2() throws Exception {
+		throws IOException, InterruptedException {
+
+		return createRemoteInputChannel(
+				inputGate, mock(PartitionRequestClient.class), 0, 0, memorySegmentProvider);
```
better avoid mock `PartitionRequestClient`.

The same issue for creating `RemoteInputChannel` in various units, if we do not need unify the `createPartitionRequestClient` with the constructing of `RemoteInputChannel`, all these `createRemoteInputChannel` could be covered by one `InputChannelBuilder`.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultPartition.java:
```diff
@@ -181,13 +188,16 @@ public ResultPartition(
-				"Bug in result partition setup logic: Buffer pool has not enough guaranteed buffers for this result partition.");
+			"Bug in result partition setup logic: Buffer pool has not enough guaranteed buffers for this result partition.");
 
 		checkState(this.bufferPool == null, "Bug in result partition setup logic: Already registered buffer pool.");
```
this check should be put before `resultPartitionBufferPoolFactory.create`

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java:
```diff
@@ -836,6 +833,21 @@ else if (transitionState(current, ExecutionState.FAILED, t)) {
 		}
 	}
 
+	private void setupPartionsAndGates() throws IOException {
```
why not remove this method to use the following static directly?

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/taskmanager/TaskTest.java:
```diff
@@ -262,7 +268,29 @@ public void testExecutionFailsInBlobsMissing() throws Exception {
 
 	@Test
-	public void testExecutionFailsInNetworkRegistration() throws Exception {
+	public void testExecutionFailsInNetworkRegistrationFprPartitions() throws Exception {
```
Fpr->For

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/PartitionBufferPoolFactory.java:
```diff
@@ -0,0 +1,31 @@
+
+import java.io.IOException;
+
+/** factory of {@link BufferPool} for network partitions or channels. */
```
This comment annotation /** */ is used for variable, and the class should be like following
/**
* 
*/
I am not sure of it.

azagrebin commented:
Thanks for the review @zhijiangW !
I addressed the smaller comments. The order of commits seems to have improved now :)

- As for `PartitionBufferPoolFactory`, it is specific for partitions and `BufferPoolFactory` is a general one. `PartitionBufferPoolFactory` can be also seen as wrapping `BufferPoolFactory` with the partition concern. I do not think we should include partition type or something to the `BufferPoolFactory`. The `PartitionBufferPoolFactory.create` basically transforms partition type and network configuration into general parameters of `BufferPoolFactory` so I think they belong to different level of abstractions. Alternatively, we could just add more config parameters to partition/gate constructors (config and `BufferPoolFactory`) and keep `PartitionBufferPoolFactory` logic inside but this would explode the constructors and partition/gate classes even more. It should not hurt to have each class to address as less concerns as possible.

- Commits for `PartitionBufferPoolFactory` factories address a certain concern, independent from setup commits. They can be compiled and tested independently as well. Not sure how temporary they are. At least, it should simplify review. We can of course squash them before merge.

- `ResultPartitionFactory` encapsulates all components which are needed to create any partition. `ResultPartitionFactory.create` accepts only partition specific parameters. Overall I think it should reduce creation of partition in place and should not hurt to have this concern separately. I will also move creation of subpartitions into the factory.

zhijiangW commented:
@azagrebin thanks for the confirmation.

As for `PartitionBufferPoolFactory`:

- The current `BufferPoolFactory` exists only for partition/gate, and no other parts would need `BufferPool` in flink stack atm. The new introduced `PartitionBufferPoolFactory` is also for partition/gate, but has no relationship with `BufferPoolFactory` in the architecture form. So I think it is not very specific for existing two parallel interfaces related with `BufferPool`.

- We could also pass the `ResultPartitionType` in constructor of `AbstractPartitionBufferPoolFactory`, then the new proposed interface method is shown as `create(int size, BufferPoolOwner owner)`, so the only difference with previous methods in `BufferPoolFactory` is we do not pass the `maxUsedBuffers` which is calculated based on `ResultPartitionType`.

- The key motivation of proposing new interface is we could only pass one parameter factory in constructor of `ResultPartition`, otherwise we need two parameters of `NetworkEnvironmentConfiguration` and `BufferPoolFactory`. But for the constructor of `SingleInputGate`, it seems the same because we could use `NetworkEnvironmentConfiguration` to replace the current `isCreditBased` parameter. 

- Another possible advantage of new interface is we extract the logic of creating `BufferPool` to another place, then it makes simple for `SingleInputGate`,`ResultPartition` and `NetworkEnvironment`.

- From the aspect of extended architecture, it seems no problem if we did nothing. Because different `ShuffleService` implementations might need different `InputGate` and `ResultPartitionWriter` implementations. Even we might extract the current `BlockingSubpartition` as a separate implementation of `ResultPartitionWriter` in future, and it does not need `BufferPoolFactory` any more.

In summary I do not think it is very necessary to introduce new interface here, or not very worthing.  If we want to make constructor of partition/gate simple, we could only pass `Supplier<BufferPool>` as you suggested before, and it could be explained by `NetworkEnvironment` itself. If the way of `Supplier` seems not good, I have another option:

- Add new method `createBufferPool(int numRequiredBuffers, BufferPoolOwner owner)` in existing `BufferPoolFactory`.

- Make `AbstractPartitionBufferPoolFactory` implements `BufferPoolFactory` and we could also pass `ResultPartitionType` into constructor of `AbstractPartitionBufferPoolFactory`.

- `ResultPartitionBufferPoolFactory/InputGateBufferPoolFactory` extends `AbstractPartitionBufferPoolFactory` and they only need to implement the new method `createBufferPool(int numRequiredBuffers, BufferPoolOwner owner)`.

To do so we extend and reuse the general `BufferPoolFactory` and only abstract/distinguish the implementation of calculating `maxPoolSize` in partition/gate. What do you think?

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java:
```diff
@@ -151,7 +159,7 @@ public static NetworkEnvironment create(
 			resultPartitionFactory,
 			singleInputGateFactory,
-			resultPartitionBufferPoolFactory);
+			resultPartitionBufferPoolFactory, inputGateBufferPoolFactory);
```
inputGateBufferPoolFactory should be in separate line for the commit of `[hotfix][network] Introduce InputGateBufferPoolFactory`

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/PartitionBufferPoolFactory.java:
```diff
@@ -0,0 +1,33 @@
+ * Factory of {@link BufferPool} for network partitions or channels.
+ */
+public interface PartitionBufferPoolFactory {
+	BufferPool create(int size, ResultPartitionType type, BufferPoolOwner owner) throws IOException;
```
empty line before `create` method?

zhijiangW commented:
Commits for `PartitionBufferPoolFactory`: In my previous review I saw it was not in the constructor of `NetworkEnvironment` in that commit, and then changed into the constructor in the following commit. So I mean the first introduced commit should be squashed with following commit to make sure it is in the constructor when introduced. Now I see it is already in right state, maybe some UI issue for commit sequence you confirmed before. :)

zhijiangW commented:
As for `ResultPartitionFactory`: The `ShuffleService` already takes the role of factory for creating partition/gate. Of course we could further extract a specific factory component from `ShuffleService`. But I think the factory might be different for different `ShuffleService` implementations, and it is regarded as a private tool inside specific `ShuffleService` implementation. 

E.g. for `FSShuffleService#FSResultPartition`, we might do not need the institute of `subpartition`, only need the logic `subpartition` index. Also for interactive queries, if it wants to reuse the `ResultPartitionWriter`, it also does not need to construct `subpartition` inside specific instance.

azagrebin commented:
Thanks for more review @zhijiangW !

I removed `PartitionBufferPoolFactory` commits and changed them to function suppliers. Normal partition/gate factories are now responsible for buffer pool creation logic. We can reintroduce `PartitionBufferPoolFactory` in future if normal factories get too many responsibilities.

True,`ShuffleService` will be a factory for partitions/gates. I just do not want it to take too many responsibilities. I think delegating further creation process to concrete factories already makes sense atm. Ideally, any class should not take more than one responsibility among couple of abstraction layers but it also depends practically on how it is already big or not, of course. Also class constructors should ideally not have any complex logic to see real class dependencies and simplify dependency injection, e.g. for testing. We can just have `BasePartitionFactory` and `FSResultPartitionFactory` for `FSResultPartition` then or see how it goes.

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultPartition.java:
```diff
@@ -154,13 +161,16 @@
+		checkArgument(bufferPool.getNumberOfRequiredMemorySegments() >= getNumberOfSubpartitions(),
+			"Bug in result partition setup logic: Buffer pool has not enough guaranteed buffers for this result partition.");
+
 		this.bufferPool = checkNotNull(bufferPool);
```
null checks is ineffective

zentol commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/LocalInputChannelTest.java:
```diff
@@ -114,11 +114,12 @@ public void testConcurrentConsumeMultiplePartitions() throws Exception {
+					networkBuffers.createBufferPool(producerBufferPoolSize, producerBufferPoolSize))
 				.build();
 
 			// Create a buffer pool for this partition
```
comment seems a bit out-dated, I'd probably just remove it

azagrebin replied:
well, this is still what actually happens atm, why is it out-dated?

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java:
```diff
@@ -88,35 +86,62 @@
+			networkBufferPool,
+			connectionManager,
+			resultPartitionManager,
+			checkNotNull(taskEventPublisher),
```
`checkNotNull` should be done in first reference with creation `SingleInputGateFactory` above.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java:
```diff
@@ -88,35 +86,62 @@
+		ResultPartitionManager resultPartitionManager = new ResultPartitionManager();
+		ResultPartitionFactory resultPartitionFactory = new ResultPartitionFactory(
+			resultPartitionManager, checkNotNull(ioManager), networkBufferPool,
+			config.networkBuffersPerChannel(), config.floatingNetworkBuffersPerGate());
```
I am not sure whether it is worth passing `config` here instead of two separate parameters like `SingleInputGateFactory`, which could avoid changing the structor of `ResultPartitionFactory` if there are more infos needed from `config` future.

azagrebin replied:
In general, I would prefer to keep classes/functions to have as least dependencies as possible if full `config` is not needed, it is easier to see the real dependencies and use in other places, like avoid mocking full config in tests.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/NetworkBufferPool.java:
```diff
@@ -69,14 +71,19 @@
 
 	private int numTotalRequiredBuffers;
 
+	private final int numberOfRequestedBuffers;
```
numberOfRequestedBuffers -> numberOfRequiredBuffers?

azagrebin replied:
I would name it `numberOfSegmentsToRequest`

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultPartitionFactory.java:
```diff
@@ -0,0 +1,199 @@
+import java.io.IOException;
+import java.util.Optional;
+
+/** Factory for {@link ResultPartition} to use in {@link org.apache.flink.runtime.io.network.NetworkEnvironment}. */
```
```
/**
*
**/
```

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultPartition.java:
```diff
@@ -125,52 +124,33 @@
 		this.partitionId = checkNotNull(partitionId);
 		this.partitionType = checkNotNull(partitionType);
-		this.subpartitions = new ResultSubpartition[numberOfSubpartitions];
+		this.subpartitions = subpartitions;
```
checkNotNull(subpartitions)?

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java:
```diff
@@ -121,24 +117,29 @@ public RemoteInputChannel(
 	 */
-	void assignExclusiveSegments(List<MemorySegment> segments) {
+	void assignExclusiveSegments() throws IOException {
+		Collection<MemorySegment> segments = memorySegmentProvider.requestMemorySegments();
```
request should be after the check of `this.initialCredit == 0` below.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGate.java:
```diff
@@ -219,6 +211,17 @@ public SingleInputGate(
 	}
 
+	@Override
+	public void setup() throws IOException {
```
In the setup, we should also `checkState(this.bufferPool == null)` firstly, then it could void bring potential trouble of releasing the related buffers if setup is called twice.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGate.java:
```diff
@@ -297,37 +300,18 @@ public void setBufferPool(BufferPool bufferPool) {
-	 * @param networkBuffersPerChannel The number of exclusive buffers for each channel
 	 */
-	public void assignExclusiveSegments(NetworkBufferPool networkBufferPool, int networkBuffersPerChannel) throws IOException {
+	public void assignExclusiveSegments() throws IOException {
```
should add @VisibleForTesting in this method.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGateFactory.java:
```diff
@@ -0,0 +1,209 @@
+		this.partitionManager = partitionManager;
+		this.taskEventPublisher = taskEventPublisher;
+		this.networkBufferPool = networkBufferPool;
+		this.networkBuffersPerChannel = networkConfig.networkBuffersPerChannel();
```
It might be better to put `this.networkBuffersPerChannel` together with above `this.partitionRequestMaxBackoff`, because these infos are all form `networkConfig`.

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/NetworkEnvironmentTest.java:
```diff
@@ -63,35 +64,31 @@
 	public void testRegisterTaskUsesBoundedBuffers() throws Exception {
 		final NetworkEnvironment network = new NetworkEnvironmentBuilder()
 			.setIsCreditBased(enableCreditBasedFlowControl)
+			.setNetworkBuffersPerChannel(2)
```
It is not necessary because the default value is 2 in `NetworkEnvironmentBuilder`

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/buffer/LocalBufferPoolDestroyTest.java:
```diff
@@ -48,7 +48,7 @@ public void testDestroyWhileBlockingRequest() throws Exception {
 
 		try {
-			networkBufferPool = new NetworkBufferPool(1, 4096);
+			networkBufferPool = new NetworkBufferPool(1, 4096, 1);
```
Maybe it is better to keep the previous constructor to avoid change many tests.
```
NetworkBufferPool(int numberOfSegmentsToAllocate, int segmentSize) {
    this (numberOfSegmentsToAllocate, segmentSize, 2)
}
```

azagrebin replied:
I did not want to explode number of constructors for tests, especially for one, not very important parameters. I think it does not make too long to add it in tests.

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/InputChannelTestUtils.java:
```diff
@@ -133,31 +103,64 @@ public static LocalInputChannel createLocalInputChannel(
+			.buildRemote(inputGate);
+	}
+
+	public static ConnectionManager mockConnectionManagerWithPartitionRequestClient(PartitionRequestClient client) {
```
Maybe it is better to construct the real `PartitionRequestClient` and return in `createPartitionRequestClient`, not pass it here because most of the usages are stilling mocking the `ConnectionID` and `PartitionRequestClientFactory` while constructing `PartitionRequestClient`.

azagrebin replied:
we could consider it, but this sounds like another refactoring :) for a later follow up

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/PartitionTestUtils.java:
```diff
@@ -40,35 +38,38 @@ public static ResultPartition createPartition(ResultPartitionType type) {
+		return createPartition(null, notifier, type, 1, sendScheduleOrUpdateConsumersMessage);
 	}
 
 	public static ResultPartition createPartition(
```
Since we already have the `ResultPartitionBuilder`, this `PartitionTestUtils` seems not very necessary because it seems trouble to maintain the different units of parameters.

azagrebin replied:
Well, certain short-cuts are still shorter if used in many places.. Not sure, I'll reduce it a bit more but atm I do not see it strictly being necessary.

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/ResultPartitionTest.java:
```diff
@@ -215,10 +215,9 @@ private void testReleaseMemory(final ResultPartitionType resultPartitionType) th
 			.setNumNetworkBuffers(numAllBuffers).build();
-		final ResultPartitionConsumableNotifier notifier = new NoOpResultPartitionConsumableNotifier();
-		final ResultPartition resultPartition = createPartition(notifier, resultPartitionType, false);
+		final ResultPartition resultPartition = createPartition(network, resultPartitionType, 1);
```
Actually the default value 1 is not needed, just because we could not give all the possible units of parameters in `PartitionTestUtils`. So it might be better to use `ResultPartitionBuilder` directly to delete intermediate `PartitionTestUtils` as I mentioned above.

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGateBuilder.java:
```diff
@@ -0,0 +1,102 @@
+			config.isCreditBased(),
+			config.networkBuffersPerChannel(),
+			config.floatingNetworkBuffersPerGate(),
+			numberOfChannels, partitionType);
```
separate line for `numberOfChannels` and `partitionType`?

zhijiangW commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/ResultPartitionBuilder.java:
```diff
@@ -0,0 +1,165 @@
+ */
+public class ResultPartitionBuilder {
+
+	private JobID jobId = new JobID();
```
should be final for these vars.

zhijiangW commented:
Thanks for the updates @azagrebin ! It seems much better now.

I have other two concerns:

- It is really nice now to introduce some builders for constructing `ResultPartition`, `SingleInputGate` and `InputChannel`. But there still exits various utils in tests to support various of parameter units for creating these items. It might bring unnecessary changes every time when make changes for partition/gate/channel. Even these scattered utils make hard to maintain and still could not provide everything. So I want to remove all these unnecessary utils in tests to unified use the proposed builder, maybe not in the scope of this PR.

- I was ever a bit confused of the proposed `MemorySegmentProvider` interface, because I think the new introduced methods should be extended in existing `BufferProvider` and `BufferRecycler` to support the batch of buffer requesting/recycle. But now I think it seems make sense to do as now, because it is very simple to let `NetworkBufferPool` implement the new interface and `NetworkBufferPool` actually has the corresponding functions internally before.

azagrebin commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/InputGateConcurrentTest.java:
```diff
@@ -88,7 +90,9 @@ public void testConsumptionWithRemoteChannels() throws Exception {
 		final ConnectionManager connManager = createDummyConnectionManager();
 		final Source[] sources = new Source[numberOfChannels];
 
-		final SingleInputGate gate = createSingleInputGate(numberOfChannels);
```
True, actually I do not mind short-hands. Though, I would remove one of short-hands with 3 parameters. I think builder is more readable in case of many parameters.

azagrebin commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGateBuilder.java:
```diff
@@ -0,0 +1,82 @@
+
+	private JobID jobId = new JobID();
+
+	private IntermediateDataSetID intermediateDataSetID = new IntermediateDataSetID();
```
though builder is not supposed to be immutable in general, we can keep it final for now and change if needed

azagrebin commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/InputChannelBuilder.java:
```diff
@@ -0,0 +1,145 @@
+			initialBackoff,
+			maxBackoff,
+			metrics);
+		inputGate.setInputChannel(partitionId.getPartitionId(), channel);
```
true, I can add it explicitly to the method name but it is how it is used in almost all tests, so I would leave it.

azagrebin commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/netty/PartitionRequestClientTest.java:
```diff
@@ -106,15 +106,15 @@ public void testDoublePartitionRequest() throws Exception {
 			channel, handler, mock(ConnectionID.class), mock(PartitionRequestClientFactory.class));
 
-		final NetworkBufferPool networkBufferPool = new NetworkBufferPool(10, 32);
+		final int numExclusiveBuffers = 2;
```
I left it where it was used somewhere else, like in assertions, for readability

azagrebin commented:
Thanks for the reviews @zentol @zhijiangW , I addressed the comments

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java:
```diff
@@ -115,7 +115,11 @@ public static NetworkEnvironment create(
 			MetricGroup metricGroup,
 			IOManager ioManager) {
-		NettyConfig nettyConfig = checkNotNull(config).nettyConfig();
+		checkNotNull(ioManager);
```
the check at L 136 is now reedundant

azagrebin replied:
true, done

azagrebin commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/InputGateFairnessTest.java:
```diff
@@ -327,7 +327,8 @@ public FairnessVerifyingInputGate(
 			super(owningTaskName, jobId, consumedResultId, ResultPartitionType.PIPELINED,
-				consumedSubpartitionIndex, numberOfInputChannels, taskActions, new SimpleCounter(), isCreditBased);
+				consumedSubpartitionIndex, numberOfInputChannels, taskActions, new SimpleCounter(),
+				isCreditBased, () -> null);
```
I would prefer `throw new UnsupportedOperationException()`. There is no need for `Optional` in production code.

pnowojski replied:
In that case `throw new UnsupportedOperationException` is of course better :)

azagrebin commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/ResultPartitionBuilder.java:
```diff
@@ -48,6 +55,14 @@
+
+	private int floatingNetworkBuffersPerGate = 1;
+
+	private FunctionWithException<BufferPoolOwner, BufferPool, IOException> bufferPoolFactory;
```
Is it related to the previous comment about `UnsupportedOperationException`?

pnowojski replied:
I think not. This builder is later performing `!= null` check on this field

azagrebin replied:
ok, I see, done

azagrebin commented:
Thanks for the review @pnowojski, I've addressed comments and left a comment about `Optional`.
